### PR TITLE
Make normals consistent

### LINKED ISSUE #17539: Use same normal field in mortar constraint objects as we use for projection when building the mortar mesh
## Reason
Currently we use a `_normals` in `(AD)MortarConstraints` that is obtained essentially through `fe_map->get_normals()`
after reinit'ing the secondary element face at reference points provided by the mortar mesh. I don't believe this results in variational consistency (it certainly doesn't result in *consistency* within the mortar system) for mechanical contact methods.

## Design
I'm not quite sure yet. Currently we get it very simply through `_assembly`. I think we might have to go through the `AutomaticMortarGeneration` object to get this right.

## Impact
Certainly more consistency in our mortar objects (e.g. use the same normal field for *everything*), and hopefully more variational consistency as well.

### PR DESCRIPTION
**Pull request by lindsayad**
Also (perhaps temporarily) adds the option to interpolate the nodal normals or use them straight (e.g. n vs n_j from the Popp literature).

Summary of some convergence results here: https://github.com/idaholab/moose/issues/17539#issuecomment-816011170

@recuero

### CODE REVIEW

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 1]
```diff
@@ -117,6 +116,55 @@ AutomaticMortarGeneration::buildNodeToElemMaps()
   }
 }
 
+std::vector<Point>
+AutomaticMortarGeneration::getNodalNormals(const Elem & secondary_elem) const
+{
+  std::vector<Point> nodal_normals(secondary_elem.n_nodes());
+  for (const auto n : make_range(secondary_elem.n_nodes()))
+    nodal_normals[n] = secondary_node_to_nodal_normal.at(secondary_elem.node_ptr(n));
```
Comment (recuero) [inline framework/src/constraints/AutomaticMortarGeneration.C:124]: Here, the normals belong to lower-dimensional elements, ie. `secondary_elem` is always a one-dimensional element?

Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:124]: Correct...this should only ever be called with a lower dimensional element from the secondary face

Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:124]: I should perhaps add an assertion that `secondary_elem.dim() == 1`

File: framework/include/constraints/AutomaticMortarGeneration.h [hunk 2]
```diff
@@ -152,7 +152,64 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
...
+   * are always 1D
+   * @return The normals
+   */
+  std::vector<Point> getNormals(const Elem & secondary_elem,
+                                const std::vector<Point> & xi1_pts) const;
+
+  /**
+   * Compute the normals at given reference points on a secondary element
+   * @param secondary_elem The secondary element used to query for associated nodal normals
+   * @param 1d_xi1_pts The reference points on the secondary element to evaluate the normals at. The
```
Comment (recuero) [inline framework/include/constraints/AutomaticMortarGeneration.h:174]: `1d` -> `oned`

Comment (recuero) [inline framework/include/constraints/AutomaticMortarGeneration.h:174]: `1d` -> `oned`

File: framework/include/constraints/AutomaticMortarGeneration.h [hunk 3]
```diff
@@ -152,7 +152,64 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
    */
   bool onDisplaced() const { return _on_displaced; }
 
-public:
+  /**
+   * @return The nodal normals associated with the provided \p secondary_elem
```
Comment (recuero) [inline framework/include/constraints/AutomaticMortarGeneration.h:156]: The comments are pretty clear, do we usually add all arguments with `@param` or is it optional?

File: framework/src/constraints/MortarConstraintBase.C [hunk 4]
```diff
@@ -68,6 +68,12 @@ MortarConstraintBase::validParams()
       "compute_primal_residuals", true, "Whether to compute residuals for the primal variable.");
   params.addParam<bool>(
       "compute_lm_residuals", true, "Whether to compute Lagrange Multiplier residuals");
+  params.addParam<bool>(
+      "interpolate_normals",
+      true,
+      "Whether to interpolate the nodal normals (e.g. classic idea of evaluating field at "
+      "quadrature points). If this is set to false, then non-interpolated nodal normals will be "
```
Comment (recuero) [inline framework/src/constraints/MortarConstraintBase.C:75]: The non-interpolated normal normals are still averaged compared to the element face normals

**Review by recuero — APPROVED** [review]
Looks great to me. Helps with convergence and results consistency 👍

File: framework/include/constraints/AutomaticMortarGeneration.h [hunk 5]
```diff
@@ -152,7 +152,64 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
...
+   */
+  std::vector<Point> getNormals(const Elem & secondary_elem,
+                                const std::vector<Real> & oned_xi1_pts) const;
+
+  /**
+   * @return The mortar interface coupling
+   */
+  const std::unordered_multimap<dof_id_type, dof_id_type> & mortarInterfaceCoupling() const
+  {
+    return mortar_interface_coupling;
```
Comment (dschwen) [inline framework/include/constraints/AutomaticMortarGeneration.h:187]: Member variables without a leading underscore. Not introduced by this PR, but maybe you'd like to jump at the opportunity to add another commit to this? :-D

File: framework/src/constraints/AutomaticMortarGeneration.C (hunk 1 above)
Comment (dschwen) [reply framework/src/constraints/AutomaticMortarGeneration.C:124]: But there is a way forward to do this for 2d secondary elements, right?

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 6]
```diff
@@ -117,6 +116,55 @@ AutomaticMortarGeneration::buildNodeToElemMaps()
...
+
+  return nodal_normals;
+}
+
+std::vector<Point>
+AutomaticMortarGeneration::getNormals(const Elem & secondary_elem,
+                                      const std::vector<Real> & oned_xi1_pts) const
+{
+  std::vector<Point> xi1_pts(oned_xi1_pts.size());
+  for (const auto qp : index_range(oned_xi1_pts))
```
Comment (dschwen) [inline framework/src/constraints/AutomaticMortarGeneration.C:134]: `index_range` looks much nicer than that macro wrapped decltype abomination!

**Review by dschwen — APPROVED** [review]

**Pull request merged**